Port WebExtensionCallbackHandler to CPP

ItsJamie9494 opened the pull request:
#### 1f6e780fe3eecf00bfd700423e6a086481a760aa
<pre>
Port WebExtensionCallbackHandler to CPP
<a href="https://bugs.webkit.org/show_bug.cgi?id=301784">https://bugs.webkit.org/show_bug.cgi?id=301784</a>

Reviewed by Timothy Hatcher.

This updates the WebExtensionCallbackHandler code to use C++, and ports all usage of call() to using JSValueRefs. As well, utility functions to create JSValueRefs from arrays, JSON, or dictionary objects are added, instead of relying on the JSValue* Cocoa API and converting that to JSValueRef.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIActionCocoa.mm:
(WebKit::WebExtensionAPIAction::getTitle):
(WebKit::WebExtensionAPIAction::getBadgeText):
(WebKit::WebExtensionAPIAction::getBadgeBackgroundColor):
(WebKit::WebExtensionAPIAction::isEnabled):
(WebKit::WebExtensionAPIAction::getPopup):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIAlarmsCocoa.mm:
(WebKit::WebExtensionAPIAlarms::get):
(WebKit::WebExtensionAPIAlarms::getAll):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPICommandsCocoa.mm:
(WebKit::WebExtensionAPICommands::getAll):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPICookiesCocoa.mm:
(WebKit::WebExtensionAPICookies::get):
(WebKit::WebExtensionAPICookies::getAll):
(WebKit::WebExtensionAPICookies::set):
(WebKit::WebExtensionAPICookies::remove):
(WebKit::WebExtensionAPICookies::getAllCookieStores):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionAPIDeclarativeNetRequest::getEnabledRulesets):
(WebKit::WebExtensionAPIDeclarativeNetRequest::getDynamicRules):
(WebKit::WebExtensionAPIDeclarativeNetRequest::getSessionRules):
(WebKit::WebExtensionAPIDeclarativeNetRequest::getMatchedRules):
(WebKit::WebExtensionAPIDeclarativeNetRequest::isRegexSupported):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDevToolsExtensionPanelCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchDevToolsExtensionPanelShownEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDevToolsInspectedWindowCocoa.mm:
(WebKit::WebExtensionAPIDevToolsInspectedWindow::eval):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDevToolsPanelsCocoa.mm:
(WebKit::WebExtensionAPIDevToolsPanels::createPanel):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
(WebKit::WebExtensionAPIEvent::invokeListenersWithArgument):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIExtensionCocoa.mm:
(WebKit::WebExtensionAPIExtension::getBackgroundPage):
(WebKit::WebExtensionAPIExtension::getViews):
(WebKit::WebExtensionAPIExtension::isAllowedFileSchemeAccess):
(WebKit::WebExtensionAPIExtension::isAllowedIncognitoAccess):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm:
(WebKit::WebExtensionAPILocalization::getAcceptLanguages):
(WebKit::WebExtensionAPILocalization::getPreferredSystemLanguages):
(WebKit::WebExtensionAPILocalization::getSystemUILanguage):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIMenusCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchMenusClickedEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
(WebKit::WebExtensionAPIPermissions::getAll):
(WebKit::WebExtensionAPIPermissions::contains):
(WebKit::WebExtensionAPIPermissions::request):
(WebKit::WebExtensionAPIPermissions::remove):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPortCocoa.mm:
(WebKit::WebExtensionAPIPort::fireMessageEventIfNeeded):
(WebKit::WebExtensionAPIPort::fireDisconnectEventIfNeeded):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIRuntime::getPlatformInfo):
(WebKit::WebExtensionAPIRuntime::getBackgroundPage):
(WebKit::WebExtensionAPIRuntime::sendMessage):
(WebKit::WebExtensionAPIRuntime::sendNativeMessage):
(WebKit::WebExtensionAPIWebPageRuntime::sendMessage):
(WebKit::WebExtensionContextProxy::internalDispatchRuntimeMessageEvent):
(WebKit::WebExtensionContextProxy::internalDispatchRuntimeConnectEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
(WebKit::WebExtensionAPIScripting::executeScript):
(WebKit::WebExtensionAPIScripting::getRegisteredContentScripts):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
(WebKit::WebExtensionAPIStorageArea::get):
(WebKit::WebExtensionAPIStorageArea::getKeys):
(WebKit::WebExtensionAPIStorageArea::getBytesInUse):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::createTab):
(WebKit::WebExtensionAPITabs::query):
(WebKit::WebExtensionAPITabs::get):
(WebKit::WebExtensionAPITabs::getCurrent):
(WebKit::WebExtensionAPITabs::getSelected):
(WebKit::WebExtensionAPITabs::duplicate):
(WebKit::WebExtensionAPITabs::update):
(WebKit::WebExtensionAPITabs::getZoom):
(WebKit::WebExtensionAPITabs::detectLanguage):
(WebKit::WebExtensionAPITabs::captureVisibleTab):
(WebKit::WebExtensionAPITabs::sendMessage):
(WebKit::WebExtensionAPITabs::executeScript):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationCocoa.mm:
(WebKit::WebExtensionAPIWebNavigation::getAllFrames):
(WebKit::WebExtensionAPIWebNavigation::getFrame):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationEventCocoa.mm:
(WebKit::WebExtensionAPIWebNavigationEvent::invokeListenersWithArgument):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestCocoa.mm:
(WebKit::WebExtensionContextProxy::resourceLoadDidSendRequest):
(WebKit::WebExtensionContextProxy::resourceLoadDidPerformHTTPRedirection):
(WebKit::WebExtensionContextProxy::resourceLoadDidReceiveResponse):
(WebKit::WebExtensionContextProxy::resourceLoadDidCompleteWithError):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestEventCocoa.mm:
(WebKit::WebExtensionAPIWebRequestEvent::invokeListenersWithArgument):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm:
(WebKit::WebExtensionAPIWindows::createWindow):
(WebKit::WebExtensionAPIWindows::get):
(WebKit::WebExtensionAPIWindows::getCurrent):
(WebKit::WebExtensionAPIWindows::getLastFocused):
(WebKit::WebExtensionAPIWindows::getAll):
(WebKit::WebExtensionAPIWindows::update):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsEventCocoa.mm:
(WebKit::WebExtensionAPIWindowsEvent::invokeListenersWithArgument):
* Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
(WebKit::WebExtensionCallbackHandler::WebExtensionCallbackHandler): Deleted.
(WebKit::WebExtensionCallbackHandler::~WebExtensionCallbackHandler): Deleted.
(WebKit::WebExtensionCallbackHandler::callbackFunction const): Deleted.
(WebKit::callWithArguments): Deleted.
(WebKit::WebExtensionCallbackHandler::reportError): Deleted.
(WebKit::WebExtensionCallbackHandler::call): Deleted.
(WebKit::toString): Deleted.
(WebKit::toJSCallbackHandler): Deleted.
(WebKit::deserializeJSONString): Deleted.
(WebKit::serializeJSObject): Deleted.
(WebKit::toJSError): Deleted.
(WebKit::toJSString): Deleted.
(WebKit::toWindowObject): Deleted.
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.cpp:
(WebKit::WebExtensionCallbackHandler::WebExtensionCallbackHandler):
(WebKit::WebExtensionCallbackHandler::~WebExtensionCallbackHandler):
(WebKit::WebExtensionCallbackHandler::callbackFunction const):
(WebKit::callWithArguments):
(WebKit::WebExtensionCallbackHandler::reportError):
(WebKit::WebExtensionCallbackHandler::call):
(WebKit::toJSCallbackHandler):
(WebKit::toString):
(WebKit::toWindowObject):
(WebKit::toJSValueRef):
(WebKit::toJSError):
(WebKit::deserializeJSONString):
(WebKit::serializeJSObject):
(WebKit::fromJSONArray):
(WebKit::fromJSONObject):
(WebKit::fromJSON):
(WebKit::fromArray):
(WebKit::fromObject):
(WebKit::isDictionary):
(WebKit::isRegularExpression):
(WebKit::isThenable):
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
(WebKit::toJSString):
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_generateImplementationFile):
(_dynamicAttributesImplementation):
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
(WebKit::WebExtensionControllerProxy::globalObjectIsAvailableForFrame):
(WebKit::WebExtensionControllerProxy::serviceWorkerGlobalObjectIsAvailableForFrame):
(WebKit::WebExtensionControllerProxy::addBindingsToWebPageFrameIfNecessary):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::enumerateFramesAndNamespaceObjects):
(WebKit::WebExtensionContextProxy::enumerateFramesAndWebPageNamespaceObjects):

Canonical link: <a href="https://commits.webkit.org/302549@main">https://commits.webkit.org/302549@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/109e2e364239931a9ded4adf4388ec8617d5b1ee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129359 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1615 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40197 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136736 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80765 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131230 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1545 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1492 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98518 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66423 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132306 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1206 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115860 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79167 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1128 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33990 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80013 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109580 "Found 1 new API test failure: TestWebKitAPI.SiteIsolation.NavigationAfterWindowOpen (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34491 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139209 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1406 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1357 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107044 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1448 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112205 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106887 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1146 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30725 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/54052 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20201 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1477 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64840 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1294 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1330 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1399 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -677,16 +676,16 @@ static bool matches(WebFrame& frame, const std::optional<WebExtensionMessageTarg
             // with a signature to translate the JS function arguments. Having the block capture
             // callbackAggregatorWrapper ensures that callbackAggregator remains in scope.
-            id returnValue = listener->call(message, senderInfo, ^(JSValue *replyMessage) {
+            JSValueRef returnValue = listener->call(toJSValueRef(listener->globalContext(), message), toJSValueRef(listener->globalContext(), senderInfo), toJSValueRef(listener->globalContext(), ^(JSValue *replyMessage) {
```
```suggestion
            auto returnValue = listener->call(toJSValueRef(listener->globalContext(), message), toJSValueRef(listener->globalContext(), senderInfo), toJSValueRef(listener->globalContext(), ^(JSValue *replyMessage) {
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -52,10 +52,15 @@
+        JSGlobalContextRef globalContext = callback->globalContext();
+
+        JSValueRef permissionsValue = fromArray(globalContext, WTFMove(permissions));
+        JSValueRef originsValue = fromArray(globalContext, WTFMove(origins));
```
```suggestion
        auto globalContext = callback->globalContext();

        auto permissionsValue = fromArray(globalContext, WTFMove(permissions));
        auto originsValue = fromArray(globalContext, WTFMove(origins));
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDevToolsInspectedWindowCocoa.mm:
```diff
@@ -78,18 +78,23 @@
 
-        auto *undefinedValue = [JSValue valueWithUndefinedInContext:[JSContext contextWithJSGlobalContextRef:callback->globalContext()]];
+        JSGlobalContextRef globalContext = callback->globalContext();
+        JSValueRef undefinedValue = JSValueMakeUndefined(globalContext);
```
```suggestion
        auto globalContext = callback->globalContext();
        auto undefinedValue = JSValueMakeUndefined(globalContext);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -201,35 +201,34 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
-    };
-
-    callback->call(platformInfo);
+    JSGlobalContextRef globalContext = callback->globalContext();
```
```suggestion
    auto globalContext = callback->globalContext();
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -201,35 +201,34 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
 
 #if PLATFORM(MAC)
-    static NSString * const osValue = @"mac";
+    static constexpr auto osValue = "mac";
```
Should use `ASCIILiteral` for these.

```suggestion
    static constexpr auto osValue = "mac"_s;
```

ItsJamie9494 replied:
Yes, I agree, but it seems to create an ambiguous call to `toJSString`. I'm wondering how useful the simple `const char*` version of that function really is...

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationEventCocoa.mm:
```diff
@@ -52,7 +52,9 @@
-        Ref { *listener.first }->call(argument);
+        Ref listenerRef = Ref { *listener.first };
+        JSGlobalContextRef globalContext = listenerRef->globalContext();
+        listenerRef->call(toJSValueRef(globalContext, argument));
```
```suggestion
        listenerRef->call(toJSValueRef(listenerRef->globalContext(), argument));
```
or
```suggestion
        auto globalContext = listenerRef->globalContext();
        listenerRef->call(toJSValueRef(globalContext, argument));
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsEventCocoa.mm:
```diff
@@ -54,7 +54,9 @@
-        Ref { *listener.first }->call(argument);
+        Ref listenerRef = Ref { *listener.first };
+        JSGlobalContextRef globalContext = listenerRef->globalContext();
+        listenerRef->call(toJSValueRef(globalContext, argument));
```
```suggestion
        listenerRef->call(toJSValueRef(listenerRef->globalContext(), argument));
```
or
```suggestion
        auto globalContext = listenerRef->globalContext();
        listenerRef->call(toJSValueRef(globalContext, argument));
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.cpp:
```diff
@@ -36,6 +36,355 @@
+        retArray.append(fromJSON(context, value.get()));
+
+    auto jsArray = JSObjectMakeArray(context, retArray.size(), retArray.span().data(), nullptr);
+    return jsArray;
```
```suggestion
    return JSObjectMakeArray(context, retArray.size(), retArray.span().data(), nullptr);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.cpp:
```diff
@@ -36,6 +36,355 @@
+        return JSValueMakeUndefined(context);
+
+    Vector<JSValueRef> retArray;
+
+    for (Ref value : array)
```
```suggestion
    Vector<JSValueRef> result;
    for (Ref value : array)
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.cpp:
```diff
@@ -36,6 +36,355 @@
+    if (!object)
+        return JSValueMakeUndefined(context);
+
+    auto jsObject = JSObjectMake(context, nullptr, nullptr);
```
```suggestion
    auto result = JSObjectMake(context, nullptr, nullptr);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.cpp:
```diff
@@ -36,6 +36,355 @@
+    if (!context)
+        return nullptr;
+
+    auto jsObject = JSObjectMake(context, nullptr, nullptr);
```
```suggestion
    auto result = JSObjectMake(context, nullptr, nullptr);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.cpp:
```diff
@@ -36,6 +36,355 @@
+    ASSERT(context);
+
+    RELEASE_LOG_ERROR(Extensions, "Exception thrown: %" PUBLIC_LOG_STRING, string.utf8().data());
+
+    JSValueRef messageArgument = toJSValueRef(context, string, NullOrEmptyString::NullStringAsEmptyString);
+    return JSObjectMakeError(context, 1, &messageArgument, nullptr);
```
```suggestion
    JSValueRef messageArgument = toJSValueRef(context, string, NullOrEmptyString::NullStringAsEmptyString);

    return JSObjectMakeError(context, 1, &messageArgument, nullptr);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.cpp:
```diff
@@ -36,6 +36,355 @@
+    case NullOrEmptyString::NullStringAsEmptyString:
+        if (JSRetainPtr stringRef = toJSString(string)) {
+            // This is a safer cpp false positive (rdar://163760990).
+            SUPPRESS_UNCOUNTED_ARG return JSValueMakeString(context, stringRef.get());
+        }
+        return JSValueMakeNull(context);
```
```suggestion
        }

        return JSValueMakeNull(context);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.cpp:
```diff
@@ -36,6 +36,355 @@
+
+    callWithArguments<1>(m_rejectFunction, m_globalContext, {
+        error
+    });
```
```suggestion
    callWithArguments<1>(m_rejectFunction, m_globalContext, { error });
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDevToolsInspectedWindowCocoa.mm:
```diff
@@ -78,18 +78,23 @@
         }
 
-        RetainPtr scriptResult = result.value()->toID();
+        auto scriptResult = result.value()->toJS(globalContext);
```
```suggestion
        JSRetainPtr scriptResult = result.value()->toJS(globalContext);
```

ItsJamie9494 replied:
toJS returns a Protected<JSValueRef> already (defined in Shared/Protected.h). Do we still need to wrap the Protected in a RetainPtr?

xeenon replied:
Oh, I was mistaken on the return type. `auto` is fine then.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -201,35 +201,34 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
 
     if (auto backgroundPage = protectedExtensionContext()->backgroundPage()) {
-        callback->call(toWindowObject(callback->globalContext(), *backgroundPage) ?: NSNull.null);
+        callback->call(toWindowObject(callback->globalContext(), *backgroundPage));
```
`toWindowObject()` should return JS `null` instead of `undefined`, since that was happening here with `NSNull.null` before.

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.cpp:
```diff
@@ -36,6 +36,348 @@
+
+    auto frameContext = frame.jsContext();
+    if (!frameContext)
+        return JSValueMakeUndefined(context);
```
```suggestion
        return JSValueMakeNull(context);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.cpp:
```diff
@@ -36,6 +36,348 @@
+    if (!frameContext)
+        return JSValueMakeUndefined(context);
+
+    return JSContextGetGlobalObject(frameContext);
```
```suggestion
    return JSContextGetGlobalObject(frameContext) ?: JSValueMakeNull(context);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.cpp:
```diff
@@ -36,6 +36,348 @@
+        if (auto value = object.getValue(key)) {
+            JSRetainPtr jsKey = toJSString(key);
+            JSValueRef jsValue = fromJSON(context, value);
+            JSObjectSetProperty(context, result, jsKey.get(), jsValue, 0, nullptr);
```
```suggestion
            auto jsValue = fromJSON(context, value);
            JSObjectSetProperty(context, result, jsKey.get(), jsValue, 0, nullptr);
```
or
```suggestion
            JSObjectSetProperty(context, result, jsKey.get(), fromJSON(context, value), 0, nullptr);
```

xeenon reviewed (changes requested):
Only a couple small things. Looks great!

ItsJamie9494 committed:
Port WebExtensionCallbackHandler to CPP
https://bugs.webkit.org/show_bug.cgi?id=301784

Reviewed by Timothy Hatcher.

This updates the WebExtensionCallbackHandler code to use C++, and ports all usage of call() to using JSValueRefs. As well, utility functions to create JSValueRefs from arrays, JSON, or dictionary objects are added, instead of relying on the JSValue* Cocoa API and converting that to JSValueRef.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIActionCocoa.mm:
(WebKit::WebExtensionAPIAction::getTitle):
(WebKit::WebExtensionAPIAction::getBadgeText):
(WebKit::WebExtensionAPIAction::getBadgeBackgroundColor):
(WebKit::WebExtensionAPIAction::isEnabled):
(WebKit::WebExtensionAPIAction::getPopup):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIAlarmsCocoa.mm:
(WebKit::WebExtensionAPIAlarms::get):
(WebKit::WebExtensionAPIAlarms::getAll):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPICommandsCocoa.mm:
(WebKit::WebExtensionAPICommands::getAll):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPICookiesCocoa.mm:
(WebKit::WebExtensionAPICookies::get):
(WebKit::WebExtensionAPICookies::getAll):
(WebKit::WebExtensionAPICookies::set):
(WebKit::WebExtensionAPICookies::remove):
(WebKit::WebExtensionAPICookies::getAllCookieStores):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionAPIDeclarativeNetRequest::getEnabledRulesets):
(WebKit::WebExtensionAPIDeclarativeNetRequest::getDynamicRules):
(WebKit::WebExtensionAPIDeclarativeNetRequest::getSessionRules):
(WebKit::WebExtensionAPIDeclarativeNetRequest::getMatchedRules):
(WebKit::WebExtensionAPIDeclarativeNetRequest::isRegexSupported):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDevToolsExtensionPanelCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchDevToolsExtensionPanelShownEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDevToolsInspectedWindowCocoa.mm:
(WebKit::WebExtensionAPIDevToolsInspectedWindow::eval):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDevToolsPanelsCocoa.mm:
(WebKit::WebExtensionAPIDevToolsPanels::createPanel):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
(WebKit::WebExtensionAPIEvent::invokeListenersWithArgument):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIExtensionCocoa.mm:
(WebKit::WebExtensionAPIExtension::getBackgroundPage):
(WebKit::WebExtensionAPIExtension::getViews):
(WebKit::WebExtensionAPIExtension::isAllowedFileSchemeAccess):
(WebKit::WebExtensionAPIExtension::isAllowedIncognitoAccess):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm:
(WebKit::WebExtensionAPILocalization::getAcceptLanguages):
(WebKit::WebExtensionAPILocalization::getPreferredSystemLanguages):
(WebKit::WebExtensionAPILocalization::getSystemUILanguage):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIMenusCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchMenusClickedEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
(WebKit::WebExtensionAPIPermissions::getAll):
(WebKit::WebExtensionAPIPermissions::contains):
(WebKit::WebExtensionAPIPermissions::request):
(WebKit::WebExtensionAPIPermissions::remove):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPortCocoa.mm:
(WebKit::WebExtensionAPIPort::fireMessageEventIfNeeded):
(WebKit::WebExtensionAPIPort::fireDisconnectEventIfNeeded):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIRuntime::getPlatformInfo):
(WebKit::WebExtensionAPIRuntime::getBackgroundPage):
(WebKit::WebExtensionAPIRuntime::sendMessage):
(WebKit::WebExtensionAPIRuntime::sendNativeMessage):
(WebKit::WebExtensionAPIWebPageRuntime::sendMessage):
(WebKit::WebExtensionContextProxy::internalDispatchRuntimeMessageEvent):
(WebKit::WebExtensionContextProxy::internalDispatchRuntimeConnectEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
(WebKit::WebExtensionAPIScripting::executeScript):
(WebKit::WebExtensionAPIScripting::getRegisteredContentScripts):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
(WebKit::WebExtensionAPIStorageArea::get):
(WebKit::WebExtensionAPIStorageArea::getKeys):
(WebKit::WebExtensionAPIStorageArea::getBytesInUse):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::createTab):
(WebKit::WebExtensionAPITabs::query):
(WebKit::WebExtensionAPITabs::get):
(WebKit::WebExtensionAPITabs::getCurrent):
(WebKit::WebExtensionAPITabs::getSelected):
(WebKit::WebExtensionAPITabs::duplicate):
(WebKit::WebExtensionAPITabs::update):
(WebKit::WebExtensionAPITabs::getZoom):
(WebKit::WebExtensionAPITabs::detectLanguage):
(WebKit::WebExtensionAPITabs::captureVisibleTab):
(WebKit::WebExtensionAPITabs::sendMessage):
(WebKit::WebExtensionAPITabs::executeScript):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationCocoa.mm:
(WebKit::WebExtensionAPIWebNavigation::getAllFrames):
(WebKit::WebExtensionAPIWebNavigation::getFrame):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationEventCocoa.mm:
(WebKit::WebExtensionAPIWebNavigationEvent::invokeListenersWithArgument):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestCocoa.mm:
(WebKit::WebExtensionContextProxy::resourceLoadDidSendRequest):
(WebKit::WebExtensionContextProxy::resourceLoadDidPerformHTTPRedirection):
(WebKit::WebExtensionContextProxy::resourceLoadDidReceiveResponse):
(WebKit::WebExtensionContextProxy::resourceLoadDidCompleteWithError):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestEventCocoa.mm:
(WebKit::WebExtensionAPIWebRequestEvent::invokeListenersWithArgument):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm:
(WebKit::WebExtensionAPIWindows::createWindow):
(WebKit::WebExtensionAPIWindows::get):
(WebKit::WebExtensionAPIWindows::getCurrent):
(WebKit::WebExtensionAPIWindows::getLastFocused):
(WebKit::WebExtensionAPIWindows::getAll):
(WebKit::WebExtensionAPIWindows::update):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsEventCocoa.mm:
(WebKit::WebExtensionAPIWindowsEvent::invokeListenersWithArgument):
* Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
(WebKit::WebExtensionCallbackHandler::WebExtensionCallbackHandler): Deleted.
(WebKit::WebExtensionCallbackHandler::~WebExtensionCallbackHandler): Deleted.
(WebKit::WebExtensionCallbackHandler::callbackFunction const): Deleted.
(WebKit::callWithArguments): Deleted.
(WebKit::WebExtensionCallbackHandler::reportError): Deleted.
(WebKit::WebExtensionCallbackHandler::call): Deleted.
(WebKit::toString): Deleted.
(WebKit::toJSCallbackHandler): Deleted.
(WebKit::deserializeJSONString): Deleted.
(WebKit::serializeJSObject): Deleted.
(WebKit::toJSError): Deleted.
(WebKit::toJSString): Deleted.
(WebKit::toWindowObject): Deleted.
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.cpp:
(WebKit::WebExtensionCallbackHandler::WebExtensionCallbackHandler):
(WebKit::WebExtensionCallbackHandler::~WebExtensionCallbackHandler):
(WebKit::WebExtensionCallbackHandler::callbackFunction const):
(WebKit::callWithArguments):
(WebKit::WebExtensionCallbackHandler::reportError):
(WebKit::WebExtensionCallbackHandler::call):
(WebKit::toJSCallbackHandler):
(WebKit::toString):
(WebKit::toWindowObject):
(WebKit::toJSValueRef):
(WebKit::toJSError):
(WebKit::deserializeJSONString):
(WebKit::serializeJSObject):
(WebKit::fromJSONArray):
(WebKit::fromJSONObject):
(WebKit::fromJSON):
(WebKit::fromArray):
(WebKit::fromObject):
(WebKit::isDictionary):
(WebKit::isRegularExpression):
(WebKit::isThenable):
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
(WebKit::toJSString):
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_generateImplementationFile):
(_dynamicAttributesImplementation):
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
(WebKit::WebExtensionControllerProxy::globalObjectIsAvailableForFrame):
(WebKit::WebExtensionControllerProxy::serviceWorkerGlobalObjectIsAvailableForFrame):
(WebKit::WebExtensionControllerProxy::addBindingsToWebPageFrameIfNecessary):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::enumerateFramesAndNamespaceObjects):
(WebKit::WebExtensionContextProxy::enumerateFramesAndWebPageNamespaceObjects):

Canonical link: https://commits.webkit.org/302549@main
